File: python3/aoc-01-02/main.py
import copy


class Solution:

    def int_code(self, integers: list) -> list:
        loops = len(integers) // 4
        if loops > 0:
            for i in range(loops):
                start = 4 * i
                op_code = integers[start]
                pos_a = integers[start + 1]
                pos_b = integers[start + 2]
                pos_c = integers[start + 3]
                if op_code == 1:
                    integers[pos_c] = integers[pos_a] + integers[pos_b]
                elif op_code == 2:
                    integers[pos_c] = integers[pos_a] * integers[pos_b]
                elif op_code == 99:
                    break
                else:
                    raise ValueError
        return integers


if __name__ == '__main__':
    solution = Solution()

    with open('input-part-1.txt') as f:
        input_part_1 = [int(x) for x in f.read().strip().split(',')]
    input_part_1[1] = 12
    input_part_1[2] = 2
    answer_1 = solution.int_code(input_part_1)
    print(f'the answer to part-1 is {answer_1[0]}')

    with open('input-part-2.txt') as f:
        input_part_2 = [int(x) for x in f.read().strip().split(',')]
    target = 19690720
    stop = False
    for pos_a in range(99):
        for pos_b in range(99):
            input_part_2_cloned = copy.copy(input_part_2)
            input_part_2_cloned[1] = pos_a
            input_part_2_cloned[2] = pos_b
            if solution.int_code(input_part_2_cloned)[0] == target:
                print(
                    f'noun is {pos_a}, verb is {pos_b}, answer to part-2 is {pos_a * 100 + pos_b}')
                stop = True
                break
        if stop:
            break

File: rust/aoc-01-02/src/main.rs
use std::error::Error;
use std::fs;

fn main() -> Result<(), Box<dyn Error>> {
    let input = fs::read_to_string("input-part-1.txt")?;
    let mut integers: Vec<i32> = input
        .split(",")
        .map(|x| x.parse::<i32>().expect("invalid integer"))
        .collect::<Vec<i32>>();
    integers[1] = 12;
    integers[2] = 2;
    let answer = int_code(&mut integers);
    println!("answer to part-1 is {}", answer[0]);

    let input = fs::read_to_string("input-part-2.txt")?;
    let integers: Vec<i32> = input
        .split(",")
        .map(|x| x.parse::<i32>().expect("invalid integer"))
        .collect::<Vec<i32>>();
    let target = 19690720;
    'outer: for pos_a in 0..99 {
        'inner: for pos_b in 0..99 {
            let mut integers = integers.clone();
            integers[1] = pos_a as i32;
            integers[2] = pos_b as i32;
            if int_code(&mut integers)[0] == target {
                println!(
                    "noun is {}, verb is {}, answer to part-2 is {}",
                    pos_a,
                    pos_b,
                    pos_a * 100 + pos_b
                );
                break 'outer;
            }
        }
    }
    Ok(())
}

fn int_code(integers: &mut Vec<i32>) -> Vec<i32> {
    match integers.len() / 4 {
        0 => integers,
        n @ _ => {
            for i in 0..n {
                let start = 4 * i;
                let op_code = integers[start] as usize;
                let pos_a = integers[start + 1] as usize;
                let pos_b = integers[start + 2] as usize;
                let pos_c = integers[start + 3] as usize;
                match op_code {
                    1 => integers[pos_c] = integers[pos_a] + integers[pos_b],
                    2 => integers[pos_c] = integers[pos_a] * integers[pos_b],
                    99 => break,
                    _ => panic!("bad op_code"),
                }
            }
            integers
        }
    }
    .to_owned()
}

#[cfg(test)]
mod tests {
    use super::int_code;

    #[test]
    fn test_int_code_part_1() {
        assert_eq!(int_code(&mut vec![1, 0, 0, 0, 99]), vec![2, 0, 0, 0, 99]);
        assert_eq!(int_code(&mut vec![2, 3, 0, 3, 99]), vec![2, 3, 0, 6, 99]);
        assert_eq!(
            int_code(&mut vec![2, 4, 4, 5, 99, 0]),
            vec![2, 4, 4, 5, 99, 9801]
        );
        assert_eq!(
            int_code(&mut vec![1, 1, 1, 4, 99, 5, 6, 0, 99]),
            vec![30, 1, 1, 4, 2, 5, 6, 0, 99]
        );
    }
}
